### ISSUE TITLE
wasm assembly type checker throws incorrect error

### ISSUE THREAD

**Issue by TerrorJack** [llvm/llvm-project#79966]
Consider the following example:

```asm
.section .text.foobar,"",@
.hidden foobar
.globl foobar
foobar:
  .functype foobar () -> (i32)
  i32.const 0
  table.get __indirect_function_table
  i32.const 42
  table.grow __indirect_function_table
  end_function
```

```c
#include <stdio.h>

int foobar(void);

int main(void) {
  printf("%d\n", foobar());
  printf("%d\n", foobar());
}
```

When compiling with `clang -mreference-types -O3 foobar.s asdf.c -o asdf.wasm -Wl,--growable-table`, it outputs the following error:

```
foobar.s:9:3: error: popped funcref, expected externref
  table.grow __indirect_function_table
  ^
```

But if I forcibly disable the wasm assembly type checker with `-Wa,--no-type-check`, then it compiles and runs just fine. The issue above is reported with latest llvm `master`.

**Comment by sbc100** [comment]
The validatotor might be correct here.  When you disabled it did  you try validating the result using something like `wasm-validate` (from wabt) to or `wasm-dis` (from binaryen), or just loading into a conforming VM?

Accroding to the spec table.grow takes and number to grow and "also takes an initialization value for the newly allocated entries".  See https://webassembly.github.io/spec/core/syntax/instructions.html#table-instructions

**Comment by TerrorJack** [comment]
I ran it with `wasmtime` and I trust `wasmtime`'s validation logic. And the first operand of `table.grow` is indeed there in `foobar`:

```asm
i32.const 0
table.get __indirect_function_table
```

The above instructions should pop a `funcref` onto the value stack.

It looks like to me that the point is: `__indirect_function_table` is supposed to be a `funcref` table, not an `externref` table, but the assembly type checker logic somehow doesn't take that into account.

**Comment by sbc100** [comment]
Ah I see.  Sounds like a bug in the validation then.

Not particularly surprising seeing as `-mreference-types` is not very heavily used yet (i.e. no frontends produce it yet).

**Comment by TerrorJack** [comment]
@HerrCai0907 Sorry, but #80437 doesn't seem to fix it, looks like the added `table.grow` typechecking logic missed an operand:

```
foobar.s:10:3: error: 1 superfluous return values
  end_function
  ^
```

**Comment by HerrCai0907** [comment]
It is not the `table.grow` issue, It is due to a known design level issue in WebAssembly backend.
A function should be keep one element in stack.

It will be valid assembly.
```
foobar:
  .functype foobar () -> (i32)
  i32.const 0
  table.get __indirect_function_table
  drop
  i32.const 42
  table.grow __indirect_function_table
  end_function
```

**Comment by TerrorJack** [comment]
@HerrCai0907 Sorry, but:

```
Error: WebAssembly translation error

Caused by:
    Invalid input WebAssembly code at offset 478: type mismatch: expected funcref but nothing on stack
```

The added type checking rule of `table.grow` in #80347 is incorrect. `table.grow` takes two operands from the value stack.

**Comment by HerrCai0907** [comment]
@TerrorJack Sorry for this omission. I have re-created a new PR to fix it.

**Comment by TerrorJack** [comment]
@HerrCai0907 Thanks for the fix 👍
